### ISSUE TITLE
iOS push notifications failing after several days

### ISSUE THREAD

**Issue by abc123s** [evollu/react-native-fcm#71]
I'm facing an issue where a new iOS user will initially be able to receive FCM push notifications, but after a few days, stops being able to receive them. However, the server which is sending the push notifications still receives a success message from FCM:

    { multicast_id: 5364950433438816000, 
      success: 1, 
      failure: 0, 
      canonical_ids: 0, 
      results: [ { message_id: '0:1471123826177313%13c7f1c613c7f1c6' } ] } 

After I tried manually sending a notification to the relevant FCM token through the firebase console (which failed), the server then starts getting a failed message from FCM:

    { multicast_id: 4711697033517732000, 
      success: 0, 
      failure: 1, 
      canonical_ids: 0, 
      results: [ { error: 'NotRegistered' } ] } 

This led me to believe that the FCM token my server had was somehow out of date; however, I double checked, and the FCM token my server is using is consistent with the token given by calling `FCM.on('refreshToken', (token) => ... ` in the app.

Is there some other reason why this might be occurring?

**Comment by evollu** [comment]
The token may change after couple days. I assume it will be through either `refreshToken` or `getFCMToken`. Are you listening on both callback?

**Comment by abc123s** [comment]
I am listening to both. The output of refreshToken still matches the FCM Token I stored on my server, but the messages still fail to send.

**Comment by evollu** [comment]
do you get refreshToken called all the time?

**Comment by abc123s** [comment]
Yes, at every app start. Often it will first receive null, but it will be called a second time with the correct token value.

**Comment by evollu** [comment]
sorry for long response time. My firebase token doesn't get refreshed often, usually on when app is reinstalled.

**Comment by sibelius** [comment]
@abc123s my guess is that there is something wrong in your code

could you post a snippet here please?

**Comment by holyxiaoxin** [comment]
I didn't know that the registration token will get refreshed. If so, it'll be quite a problem. The problem can be solved when the user logs in and refreshes the token. But what if the user is inactive for a few months and the token is refreshed in fcm side, the user can't receive any notifications then.

**Comment by evollu** [comment]
gcm token refreshes too and I think they have similar behavior. the token should remain unchanged until user log in again, but that is just my guess

**Comment by sibelius** [comment]
@evollu I think u can close this

**Comment by carlosfilipe** [comment]
Hello everyone, i have exactly the same abc123s's issue. 
I used the PushController component that is showed on sample code and I call it  on main scene of my project.
I call FCM.getFCMToken() every time that a i want retrieve the Fcm token. This method call returns the refreshed token, isn't it?

What did i may get wrong?

**Comment by evollu** [comment]
@carlosfilipe what issue are you having?

**Comment by carlosfilipe** [comment]
A new iOS user will initially be able to receive FCM push notifications, but after a few days, stops being able to receive them.

**Comment by evollu** [comment]
do you check FCM.getFCMToken() on every app launch?

**Comment by carlosfilipe** [comment]
Yes!

The app starts on main scene and a I have the follow code:
```


		<View style={{ flex: 1 }}>
                    <PushController
                      firebase={this.state.configFirebase}
                      navigator={navigator}
                      onChangeToken={token => {
                        // console.log(`onChangeToken  ${token} `);
                        this.onChangeFCMToken(token);
                      }}
                    />
                    <ListScene
                      navigator={navigator}
                      firebase={this.state.configFirebase}
                      data={route.data}
                    />
                  </View>
```

PushController component is exactly like https://github.com/evollu/react-native-fcm/blob/master/Examples/simple-fcm-client/app/PushController.js

I have a component that is responsible for user configuration. On willmount, I call :
```
FCM.getFCMToken().then(token => {
      this.fcmToken = token;
    });
```
Then i use this.fcmToken object to save user config and the token on my server.

This only happen with one of our test iOS device ( we have two). 
We observed that firebase user in this device, associated with this fcm token, had been excluded. But there is some relationship between firebase users and fcm? I think not...

(sorry about my english xD)

**Comment by evollu** [comment]
might be. If that user is no long associated with firebase, he won't get firebase token (it is the same as fcm token)

**Comment by carlosfilipe** [comment]
He keeps the same token… I think that is a firebase bug. We works with anonymous users. When we remove a firebase user by firebase console, he keeps signed in on app, with the previous token. So, if we only deactivate the user, the app verify that he is no signed, and create a new anonymous  user, with a new firebase token. And more, if we deactivate a user, and then remove it, when we start the app, firebase auth verify that users keeps signed in... a bit weird...

But a think that we don’t need use the firebase auth api to use fcm api, do we?

**Comment by evollu** [comment]
try listen on token refresh event and see if it returns anything.
otherwise I don't have any solution because it seems to be related to Firebase SDK itself

**Comment by emretekince** [comment]
i have same problem :/

**Comment by wsun** [comment]
In case this is helpful for others - I experienced a similar issue, where the FCM token had not changed, but at some point the user was no longer able to receive remote push notifications. This was because we failed to regularly keep the APNs device token up to date: we did not call `registerForRemoteNotifications` appropriately on every app launch (i.e. in AppDelegate), as Apple [requires](https://developer.apple.com/library/content/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/HandlingRemoteNotifications.html#//apple_ref/doc/uid/TP40008194-CH6-SW1). This method then triggers the Firebase-swizzled `didRegisterForRemoteNotificationsWithDeviceToken` to map the new APNs device token to the current FCM token.

This issue only seemed to occur on devices where we were regularly installing new app builds, so I suspect the APNs device token changes under certain circumstances, even when the FCM token is still the same.
